File: leetcode-124.py
# Definition for a binary tree node.
from typing import AnyStr


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    ans = -int(1e9)
    def maxPathSum(self, root: TreeNode) -> int:
        def dfs(u):
            if u is None:
                return 0
            lv = max(0, dfs(u.left))
            rv = max(0, dfs(u.right))
            self.ans = max(self.ans, lv+u.val+rv)
            return u.val+max(lv, rv)
        dfs(root)
        return self.ans

File: leetcode/leetcode-124.cpp
/*
 *@File Name          :leetcode-124.cpp
 *@Author             :cc
 *@Version            :1.0.0
 *@Date               :2022/3/3 9:23
 *@Description        :
 *@Function List      :
 *@History            :
 */

#include "algorithm"
#include "vector"

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode() : val(0), left(nullptr), right(nullptr) {}

    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}

    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int dfs(TreeNode *cur, int &ans) {
        if (cur == nullptr) return 0;
        int l = dfs(cur->left, ans);
        int r = dfs(cur->right, ans);
        if (l + r + cur->val > ans) ans = l + r + cur->val;
        if (l + cur->val > ans) ans = l + cur->val;
        if (r + cur->val > ans) ans = r + cur->val;
        if (cur->val > ans) ans = cur->val;
        return max(max(l, r), 0) + cur->val;
    }

    int maxPathSum(TreeNode *root) {
        int ans = INT_MIN;
        dfs(root, ans);
        return ans;
    }
};